sklearn/tree/_tree.pyx

```diff
@@ -114,6 +121,15 @@ cdef class TreeBuilder:
         return X, y, sample_weight
 
 # Depth first builder ---------------------------------------------------------
+# A record on the stack for depth-first tree growing
+cdef struct StackRecord:
+    SIZE_t start
+    SIZE_t end
+    SIZE_t depth
+    SIZE_t parent
+    bint is_left
+    double impurity
+    SIZE_t n_constant_features
```

Comment:
This is a naive question: is the current order of the struct fields giving the best struct memory layout (i.e. the one necessitating the less memory)?

Reply:
I probably miss the background here. If you create a structure you will zero-initialize all fields and thus memory-wise, the order would not matter?

Reply:
To make my comment more precise, I was talking about [structure packing](http://www.catb.org/esr/structure-packing/).

I've tried different definitions and in fact the size of the struct is identical in each case. Cython when compiling down to C++ or C does not change the order nor does it add, hence I suppose C/C++ compilers do it natively now?

<details>

```cython
import numpy as np
cimport numpy as np
from libc.stdio cimport printf

ctypedef np.npy_intp SIZE_t

cdef struct StackRecord_1:
    SIZE_t start
    SIZE_t end
    SIZE_t depth
    SIZE_t parent
    bint is_left
    double impurity
    SIZE_t n_constant_features


cdef struct StackRecord_2:
    SIZE_t start
    SIZE_t end
    SIZE_t depth
    SIZE_t parent
    bint is_left
    SIZE_t n_constant_features
    double impurity

cdef struct StackRecord_3:
    SIZE_t start
    bint is_left
    SIZE_t end
    double impurity
    SIZE_t depth
    SIZE_t parent
    SIZE_t n_constant_features


cdef struct StackRecord_4:
    double impurity
    bint is_left
    SIZE_t start
    SIZE_t end
    SIZE_t depth
    SIZE_t parent
    SIZE_t n_constant_features


cpdef int main() nogil:
    printf("sizeof(StackRecord_1) = %d\n", sizeof(StackRecord_1))
    printf("sizeof(StackRecord_2) = %d\n", sizeof(StackRecord_2))
    printf("sizeof(StackRecord_3) = %d\n", sizeof(StackRecord_3))
    printf("sizeof(StackRecord_4) = %d\n", sizeof(StackRecord_4))
```

gives:

```
>>> import cython_test
>>> cython_test.main()
sizeof(StackRecord_1) = 56
sizeof(StackRecord_2) = 56
sizeof(StackRecord_3) = 56
sizeof(StackRecord_4) = 56
```

</details>

Reply:
Thanks for the details.

Reply:
I suppose we can always change our minds later. scikit-learn does not serialize stack records as attributes of estimators, right? They are just used as transient datastructure used as local variables in fit or predict methods I think.

Reply:
Yes.

Actually, my initial remark can be ignored: the order has not impact for this struct (see the details in the collapsible context given above).